Added documentation for validator.md (#1161)
rossvideo/Catena: docs/Validation.md

The Catena specification is currently in discussion with OSA with the goal of SMPTE standardization. You can learn more about ST2138-a at:

- [SMPTE ST 2138-a GitHub Repository](https://github.com/SMPTE/st2138-a)

The standard defines a comprehensive device model structure using protocol buffers and JSON schemas. Device models must conform to these schemas to ensure interoperability between different implementations and manufacturers.

---

## Why a Validator

Device models can quickly become complex, spanning multiple files with hundreds of parameters, constraints, commands, and menu structures. Manual verification of these models is error-prone and time-consuming. A validator is essential for several reasons:

### 1. **Catch Errors Early**

The validator ensures your device model conforms to the ST2138-a schema before attempting to generate code or deploy the model. This catches:

- Type mismatches (e.g., using a string where an integer is required)
- Missing required fields (e.g., mandatory product information)
- Invalid enum values
- Incorrect nested structures
- Schema constraint violations

> ![Device Model Error Highlighting](images/device.minimal.error.png)

Reviewer: maybe have some example import paths? this implies to me that there exists a directory called include, and it somehow import the files in that dir.

![Auto-completion in VSCode](images/autocomplete.png)

- **Tooltips**: Shows documentation for each field

![Tooltip Documentation](images/tooltip.png)

- **Inline error highlighting**: Marks invalid syntax immediately

### 4. **Prevent Runtime Errors**

Many issues that would cause runtime failures are caught at validation time:

- Invalid JSON pointer references
- Circular dependencies in imports
- Authorization scope misconfigurations
- Constraint reference errors

The validation logic is integrated into the code generation pipeline in [index.js](../tools/codegen/index.js#L173-L187), ensuring no invalid models make it to production.

---

## Required Parts for a Valid Device Model

According to the ST2138-a specification and enforced by the [validation logic](../tools/codegen/index.js#L80-L167), a valid device model must include:

### 1. **Device Metadata**

- **`slot`**: Device slot number
- **`default_scope`**: Default access scope (e.g., `"st2138:mon"`)
- **`detail_level`**: Default detail level for responses
- **`access_scopes`**: Array of supported access scopes
- **`multi_set_enabled`**: Boolean indicating multi-set support

> - **`subscriptions`**: Boolean indicating subscription support

Reviewer: None of these are a must. Also missing fields like params, commands, languages, etc.

### 2. **Mandatory Product Parameters**

The `product` parameter must be a read-only STRUCT containing:

| Parameter | Type | Required | Description |
|-----------|------|----------|-------------|
| `name` | STRING | Yes | Product name |
| `vendor` | STRING | Yes | Manufacturer name |
| `version` | STRING | Yes | Firmware/software version |
| `serial_number` | STRING | Yes | Unique device serial number |
| `catena_sdk` | STRING | No | SDK implementation name |
| `catena_sdk_version` | STRING | No | SDK version |

All product parameters must have an access scope of `"st2138:mon"` (monitor) or inherit it from parent scopes.

Example from [use_constraints](../sdks/cpp/common/examples/use_constraints/device.use_constraints.json):

```json
{
  "slot": 1,
  "default_scope": "st2138:mon",
  "params": {
    "product": {
      "type": "STRUCT",
      "read_only": true,
      "access_scope": "st2138:mon",
      "params": {
        "name": { "type": "STRING" },
        "vendor": { "type": "STRING" },
        "version": { "type": "STRING" },

>         "serial_number": { "type": "STRING" }

Reviewer: All 6 params still have to be defined, just the optional one(s) don't need values. Should probably include an explaination about that: the struct needs all the fields, but the values can be generated automatically for them if they're optional. (looks like we're missing the generation for catena_sdk, making a ticket)

      },
      "value": {
        "struct_value": {
          "fields": {
            "name": { "string_value": "My Device" },
            "vendor": { "string_value": "Ross Video" },
            "version": { "string_value": "1.0.0" },
            "serial_number": { "string_value": "12345" }
          }
        }
      }
    }
  }
}
```

### 3. **Valid Parameter Types**

All parameters must use one of the supported types defined in the protobuf specification:

- Scalar: `INT32`, `INT64`, `FLOAT32`, `FLOAT64`, `STRING`, `BOOLEAN`
- Complex: `STRUCT`, `VARIANT`, `ARRAY`

> - Special: Commands and external objects

Reviewer: not correct, we don't have 64 types, or bool
the complex list is longer as well

See [param.proto](../smpte/interface/proto/param.proto) for complete type definitions.

### 4. **Proper File Naming Convention**

Files must follow the naming pattern for schema detection:

- Device models: `device.<name>.json` or `device.<name>.yaml`
- Parameter imports: `param.<name>.json` or `param.<name>.yaml`
- Constraint definitions: Follow schema requirements

The [validator](../smpte/tools/validator.js#L56-L84) extracts the schema name from the first part of the filename.

### 5. **Valid References**

All references must resolve correctly:

- Constraint references (`constraint_ref`)
- Template references
- Import paths
- JSON pointers in commands

---

## How to Run the Validator

There are multiple ways to validate your device models:

### Method 1: Command-Line Validation (Standalone)

The standalone validator is located in [smpte/tools/validate.js](../smpte/tools/validate.js).

**Usage:**

```bash

> cd /home/jasonchen/Catena

Reviewer: ```suggestion
cd ~/Catena
```

> node smpte/tools/validate.js <path-to-your-file>

Reviewer: should we include stuff about running smpte's install-tooling.sh and build-openapi.sh?

# Output:
# Applying schema 'device' to file './sdks/cpp/common/examples/hello_world/device.hello_world.json'
# ✅ Validation succeeded.
```

Validate a parameter file:
```bash
node smpte/tools/validate.js ./example_device_models/device_tree/params/variantlist_example/param.AudioSlot.json
# Output:
# Applying schema 'param' to file './example_device_models/device_tree/params/variantlist_example/param.AudioSlot.json'
# ✅ Validation succeeded.
```

Validate a YAML file:
```bash
node smpte/tools/validate.js ./sdks/cpp/connections/gRPC/examples/one_of_everything/device.one_of_everything.yaml
```

**How it works:**

The validator (implemented in [validator.js](../smpte/tools/validator.js)):

1. Extracts the schema name from the filename prefix (e.g., `device`, `param`)
2. Loads the appropriate schema from [device.json](../smpte/interface/schemata/device.json)
3. Parses the input file (supports both JSON and YAML)
4. Uses [AJV (Another JSON Validator)](https://ajv.js.org/) to validate against the schema
5. Reports errors with line numbers using source maps

### Method 2: NPM Package (Published Validator)

> Install the official ST2138-a validator globally:

Reviewer: It's not installable globally like that, there's no published package yet. It can be installed with a `path:` uri in the package.json, but even then you can to use npx

```bash
npm install -g smpte-st2138-a-validator
```

Then run:

```bash
st2138-a-validate ./path/to/device.good.yaml
```

Or use npx without installation:

```bash
npx smpte-st2138-a-validator ./path/to/device.good.yaml
```

This uses the published schema from the SMPTE GitHub repository.

### Method 3: Integrated Code Generation

When generating code, validation happens automatically. The [codegen tool](../tools/codegen/index.js) validates before generating:

```bash

> cd /home/jasonchen/Catena/tools/codegen

Reviewer: ```suggestion
cd ~/Catena/tools/codegen
```
best not hardcode with your username

- The root device model file
- All imported parameter files

The following device model shows two imported parameters. Full example can be found in [device.import_params.json](../sdks/cpp/common/examples/import_params/device.import_params.json):

```json
{
  "params": {
    "city": {
      "import": {"file": "params/param.city.json"}
    },
    "plane_ticket": {
      "type": "STRUCT",
      "params": {
        "departure": {
          "import": {"file": "params/param.ottawa.json"}
        },
        "destination": {
          "type": "STRUCT",
          "template_oid": "/city"
        }
      },
    }
  }
}
```

The import structure follows these rules (implemented in [DeviceModel.js](../tools/codegen/DeviceModel.js#L55-L90)):

- Root device model: Look in the `params` sub-folder for files named `param.<name>.json` or `param.<name>.yaml`
- Nested imports: Look in sub-folders matching the parent param name

File system for `device.import_params.json` example:

![Device Tree Folder Structure](images/import_params_tree.png)

### 3. **IDE Integration**

> With proper configuration, validators provide real-time feedback in your IDE:

Reviewer 2: how is the configuration setup

- **Auto-completion**: Suggests valid properties and values

![Auto-completion in VSCode](images/autocomplete.png)

- **Tooltips**: Shows documentation for each field

![Tooltip Documentation](images/tooltip.png)

- **Inline error highlighting**: Marks invalid syntax immediately

### 4. **Prevent Runtime Errors**

Many issues that would cause runtime failures are caught at validation time:

- Invalid JSON pointer references
- Circular dependencies in imports
- Authorization scope misconfigurations
- Constraint reference errors

The validation logic is integrated into the code generation pipeline in [index.js](../tools/codegen/index.js#L173-L187), ensuring no invalid models make it to production.

---

## Required Parts for a Valid Device Model

> According to the ST2138-a specification and enforced by the [validation logic](../tools/codegen/index.js#L80-L167), a valid device model could include:

Reviewer 2: ```suggestion
According to the ST2138-a specification and enforced by the [validation logic](../tools/codegen/index.js#L80-L167), a valid device model can only include:
```
